host-go/runtimes/wazero/runtime.go

```diff
@@ -16,41 +16,56 @@ import (
 )
 
 type wRuntime struct {
-	runtime wazero.Runtime
+	compilationCache wazero.CompilationCache
 }
 
 var _ module.Runtime = (*wRuntime)(nil)
 
+// New creates a new wazero wasm runtime.
+//
+// WARNING: This runtime does not current allow for instance reuse within a single pipeline.
+// Please see https://github.com/lens-vm/lens/issues/71 for more info.
 func New() module.Runtime {
-	ctx := context.TODO()
 	return &wRuntime{
-		runtime: wazero.NewRuntime(ctx),
+		compilationCache: wazero.NewCompilationCache(),
 	}
 }
 
 type wModule struct {
-	runtime wazero.Runtime
-	module  wazero.CompiledModule
+	compilationCache wazero.CompilationCache
+	moduleBytes      []byte
 }
 
 var _ module.Module = (*wModule)(nil)
 
 func (rt *wRuntime) NewModule(wasmBytes []byte) (module.Module, error) {
-	ctx := context.TODO()
-	compiledWasm, err := rt.runtime.CompileModule(ctx, wasmBytes)
-	if err != nil {
-		return nil, err
-	}
-
 	return &wModule{
-		runtime: rt.runtime,
-		module:  compiledWasm,
+		compilationCache: rt.compilationCache,
+		moduleBytes:      wasmBytes,
 	}, nil
 }
 
 func (m *wModule) NewInstance(functionName string, paramSets ...map[string]any) (module.Instance, error) {
 	ctx := context.TODO()
-	instance, err := m.runtime.InstantiateModule(ctx, m.module, wazero.NewModuleConfig().WithName(""))
+	runtimeConfig := wazero.NewRuntimeConfig().WithCompilationCache(m.compilationCache)
+	runtime := wazero.NewRuntimeWithConfig(ctx, runtimeConfig)
+
+	var nextFunction = func() module.MemSize { return 0 }
+	_, err := runtime.NewHostModuleBuilder("lens").
+		NewFunctionBuilder().
+		WithFunc(func(ctx context.Context) module.MemSize {
+			return nextFunction()
+		}).
+		Export("next").
+		Instantiate(ctx)
+	if err != nil {
+		return module.Instance{}, err
+	}
+
+	instance, err := runtime.Instantiate(ctx, m.moduleBytes)
+	if err != nil {
+		return module.Instance{}, err
+	}
```

Comment:
Does this take advantage of the per-instance optimization? It seems registering the `next` function on the runtime would prohibit multiple instances of a single compiled module, no?

Reply:
The key line here is `runtimeConfig := wazero.NewRuntimeConfig().WithCompilationCache(m.compilationCache)` - this *should* mean that the wasm module only gets compiled once.  I've not done any performance testing of this though due to https://github.com/lens-vm/lens/issues/71

Reply:
> resultant lack of interest in wazero.

I do think wazero is important due to its pure go nature. CGO has untold influence on running go programs that use it. Higher mem, higher cpu, more strain on the GC, etc.

Reply:
Yes it is important, but it is still relatively unimportant.  We as a team even agreed to drop it entirely from Lens, the fact that it is half working at all at the moment in this PR is kind of a surprise/bonus.

Reply:
As I understood it, the intention to drop was only in relation to the instance cache optimization. With the assumption that if we couldn't do it at all then we can drop wazero support until the runtime could support it.

But, with your current work around its sufficient for now.

In the long term, wazero is very important in my eyes, again to avoid *any* CGO involvement, which some ppl are rather picky/sensitive about.